CryptomatorFileProvider/Workflow/WorkflowDependency.swift

```diff
@@ -0,0 +1,56 @@
+//
+//  WorkflowDependency.swift
+//  CryptomatorFileProvider
+//
+//  Created by Philipp Schmid on 23.03.22.
+//  Copyright © 2022 Skymatic GmbH. All rights reserved.
+//
+
+import Foundation
+import Promises
+
+class WorkflowDependency<T>: WorkflowMiddleware {
+	private var next: AnyWorkflowMiddleware<T>?
+	private let parentTask: Promise<Void>
+	private let dependencies: [WorkflowDependencyNode]
+
+	/**
+	 Initializes a new workflow dependency middleware.
+
+	 - Parameter dependencies: The workflow dependency (leaf) nodes which belong to this workflow.
+	 The next middleware is executed as soon as a lock is acquired for each dependency.
+	 All leaf nodes are unlocked once the `CloudTask` associated with the workflow is completed (either successfully or with an error).
+	 */
+	init(dependencies: [WorkflowDependencyNode]) {
+		self.dependencies = dependencies
+		self.parentTask = all(dependencies.map { $0.isLocked }).then { _ -> Void in }
+	}
+
+	func setNext(_ next: AnyWorkflowMiddleware<T>) {
+		self.next = next
+	}
+
+	func getNext() throws -> AnyWorkflowMiddleware<T> {
+		guard let nextMiddleware = next else {
+			throw WorkflowMiddlewareError.missingMiddleware
+		}
+		return nextMiddleware
+	}
+
+	func execute(task: CloudTask) -> Promise<T> {
+		let nextMiddleware: AnyWorkflowMiddleware<T>
+		do {
+			nextMiddleware = try getNext()
+		} catch {
+			return Promise(error)
+		}
+		return parentTask.then {
+			nextMiddleware.execute(task: task)
+		}.then { value -> T in
+			self.dependencies.forEach { $0.unlock() }
+			return value
+		}.catch { error in
+			self.dependencies.forEach { $0.unlock(with: error) }
+		}
```

Comment:
shouldn't we unlock in reverse order? Otherwise a concurrent task may get hold of the root node and runs into a locked child.

Reply:
> I'm not sure I get the "unlock" part right. Shouldn't there be two vertical dependency chains? root-to-leaf for locking and leaf-to-root for unlocking? The latter being essential for concurrent tasks to wait on.
> 
> In general, I believe this is too complex right now. Maybe a refactoring of the current locking scheme (before this PR) would simplify things. Due to the already very complex base, much is done in a procedural way, where I doubt that promises can unfold their full potential.

Exactly there should be two vertical dependency chains (although one could argue it can be seen as a single chain if the leaf unlock "depends" on the leaf lock). I think the perhaps confusing thing is that the current structure is as follows:

I have "combined" in a workflow dependency node one lock node and the corresponding unlock node. This means that for a path component we always have a lock (marked with orange in the figure) and an unlock (marked with blue in the figure).

The dependency chain:
<img width="392" alt="DependencyChain" src="https://user-images.githubusercontent.com/25935690/160079192-27f4c2e4-ec41-484a-949d-b4851940e6c7.png">

A single workflow dependency node:
<img width="304" alt="SingleWorkflowDependencyNode" src="https://user-images.githubusercontent.com/25935690/160078934-6b55045a-3467-429b-ab49-18e0615acb20.png">

I think the most confusing part of this might be the name "workflow dependency node", since at conception we always had a node that represented either a lock or an unlock and never both. However, I think these two individual nodes are closely related semantically.